### PR TITLE
Discard optional paren history based on params evaluation

### PR DESCRIPTION
**Pull request by silasb**
Working on a URL generator based on a route path.  Using `formatPattern` gets close, but fails when we have routes like `search(/collection/:collectionId)`.  This PR takes care of that case.

### CODE REVIEW

**Comment by silasb** [comment]
Thanks for letting me know. Travis didn't notify me.

—Silas

File: modules/PatternUtils.js [hunk 1]
```diff
@@ -180,12 +187,38 @@ export function formatPattern(pattern, params) {
...
+      if (parenCount)
+        history[parenCount - 1] += token
+      else
+        pathname += token
     }
   }
 
+  invariant(
+    history.length == 0,
+    'History is not empty [%s]',
```
Comment (timdorr) [inline modules/PatternUtils.js]: This error message doesn't tell me what's going on here.

File: modules/PatternUtils.js [hunk 2]
```diff
@@ -149,7 +149,7 @@ export function formatPattern(pattern, params) {
   params = params || {}
 
   const { tokens } = compilePattern(pattern)
-  let parenCount = 0, pathname = '', splatIndex = 0
+  let parenCount = 0, pathname = '', splatIndex = 0, history = []
```
Comment (timdorr) [inline modules/PatternUtils.js]: I really wouldn't use a variable named `history`, given that this library uses that name extensively in various places.

**Comment by silasb** [comment]
Addressed the above comments.  Thanks for the feedback.

**Comment by silasb** [comment]
Any thoughts on this?

File: modules/PatternUtils.js [hunk 3]
```diff
@@ -167,23 +167,60 @@ export function formatPattern(pattern, params) {
...
+      else
+        pathname += parenText
     } else if (token.charAt(0) === ':') {
       paramName = token.substring(1)
       paramValue = params[paramName]
 
       invariant(
         paramValue != null || parenCount > 0,
-        'Missing "%s" parameter for path "%s"',
+        'missing "%s" parameter for path "%s"',
```
Comment (timdorr) [inline modules/PatternUtils.js]: Don't downcase this. It is going to be shown to a human as an error and should read as a proper sentence.

File: modules/PatternUtils.js [hunk 4]
```diff
@@ -167,23 +167,60 @@ export function formatPattern(pattern, params) {
...
+          for (let i = 0; i < tokensSubset.length; i++) {
+            if (tokensSubset[i] == ')') {
+              nextParenIdx = i
+              break
+            }
+          }
+
+          invariant(
+            nextParenIdx > 0,
+            'missing end parameter'
```
Comment (timdorr) [inline modules/PatternUtils.js]: This is still unfriendly to the end-user. I would include some context to point them in the right direction. E.g., `'Path "%s" is missing end parameter at segment "%s"', pattern, tokensSubset.join('')`

**Pull request merged**